funsor/numpy.py

```diff
@@ -125,36 +131,60 @@ def _cholesky_inverse(x):
     """
     Like :func:`torch.cholesky_inverse` but supports batching and gradients.
     """
-    return _cholesky_solve(np.eye(x.shape[-1]), x)
+    return _cholesky_solve(_new_eye(x, x.shape[:-1]), x)
 
 
 @ops.cholesky_solve.register(array, array)
 def _cholesky_solve(x, y):
-    from scipy.linalg import cho_solve
-
-    # TODO: remove this logic when using JAX
-    # work around the issue of scipy which does not support batched input
-    batch_shape = np.broadcast(x[..., 0, 0], y[..., 0, 0]).shape
-    xs = np.broadcast_to(x, batch_shape + x.shape[-2:]).reshape((-1,) + x.shape[-2:])
-    ys = np.broadcast_to(y, batch_shape + y.shape[-2:]).reshape((-1,) + y.shape[-2:])
-    ans = [cho_solve((y, True), x) for (x, y) in zip(xs, ys)]
-    ans = np.stack(ans)
-    return ans.reshape(batch_shape + ans.shape[-2:])
+    return cho_solve((y, True), x)
 
 
 @ops.triangular_solve_op.register(array, array, bool, bool)
 def _triangular_solve(x, y, upper, transpose):
-    from scipy.linalg import solve_triangular
-
-    # TODO: remove this logic when using JAX
-    # work around the issue of scipy which does not support batched input
-    batch_shape = np.broadcast(x[..., 0, 0], y[..., 0, 0]).shape
-    xs = np.broadcast_to(x, batch_shape + x.shape[-2:]).reshape((-1,) + x.shape[-2:])
-    ys = np.broadcast_to(y, batch_shape + y.shape[-2:]).reshape((-1,) + y.shape[-2:])
-    ans = [solve_triangular(y, x, trans=int(transpose), lower=not upper)
-           for (x, y) in zip(xs, ys)]
-    ans = np.stack(ans)
-    return ans.reshape(batch_shape + ans.shape[-2:])
+    assert np.ndim(x) >= 2 and np.ndim(y) >= 2
+    n, m = x.shape[-2:]
+    assert y.shape[-2:] == (n, n)
+    # NB: JAX requires x and y have the same batch_shape
+    batch_shape = lax.broadcast_shapes(x.shape[:-2], y.shape[:-2])
+    x = np.broadcast_to(x, batch_shape + (n, m))
+
+    # The following procedure handles the case: y.shape = (i, 1, n, n), x.shape = (..., i, j, n, m)
+    # because we don't want to broadcast y to the shape (i, j, n, n).
```

Comment:
This is to save computation. In principle, with A.shape = (3, 3), b.shape=(100, 3, 1), solving a single linear equation `Ax = b.swapaxis(0, 1).reshape(3, 100)` would be much faster than solving 100 linear equations `Ax_i = b[i]` for `i in range(100)`.

Reply:
I'll move this logic upstream.